iOS migration (#493)
Amplitude-Developer-Docs/amplitude-dev-center: docs/data/sdks/ios-swift/migration.md

---
title: iOS SDK Migration Guide
description: Use this guide to easily migrate from Amplitude's legacy iOS SDK (Amplitude-iOS) to the new SDK (Amplitude-Swift).
---

The new version of Amplitude's iOS SDK (`Amplitude-Swift`) features a plugin architecture, built-in type definition and broader support for front-end frameworks. The new version isn't backwards compatible with `Amplitude-iOS`. 

To migrate to `Amplitude-Swift`, update your dependencies and instrumentation.

## Terminology

* `Amplitude-iOS`: Legacy iOS SDK
* `Amplitude-Swift`: New iOS SDK

## Dependency

For CocoaPods installation:
=== "Amplitude-iOS"
    Add dependency to Podfile.
    ```
    pod 'Amplitude', '~> 8.8.0'
    ```

=== "Amplitude-Swift"
    Add dependency to Podfile.
    ```

>     pod 'AmplitudeSwift', '~> 0.3.0'

Reviewer: Albeit confusing, I feel like we should name the new pod "AmplitudeIOS" or "AmplitudeAnalytics" to avoid confusion on supported languages like we did with the "TypeScript" SDK (i.e. does this work for ObjC?)

Author: Thanks for the review. Yea, I think that's a good suggestion. I like the AmplitudeAnalytics more. But we already have a release before. Is there any concern if we change the package name now?

Author: Merge this. And will have offline discussion related to updating the library name.

    ```

=== "Amplitude-Swift"

    ```swift
    let identify = Identify()
    identify.set(property: "membership", value: "paid")
    amplitude.groupIdentify(
      groupType: "TEST-GROUP-TYPE", 
      groupName: "TEST-GROUP-NAME", 
      identify: identify
    )
    ```

### Tracking revenue

#### `logRevenueV2()`

Track revenue using `revenue()` API on `amplitude` without calling `instance()`.

=== "Amplitude-iOS"

    ```swift
      let revenue = AMPRevenue()
      revenue.setProductIdentifier("productIdentifier")
      revenue.setQuantity(3)
      revenue.setPrice(NSNumber(value: 3.99))

      Amplitude.instance().logRevenueV2(revenue)
    ```

=== "Amplitude-Swift"

    ```swift
    let revenue = Revenue()
    revenue.productId = "123"
    revenue1.price = 12
            
    amplitude.revenue(revenue: revenue)
    ```

### Patterns

#### Plugins

The configs `amplitude.adSupportBlock` or `amplitude.useAdvertisingIdForDeviceId` were available in `Amplitude-iOS` to allow you to use IDFV or IDFA as the deviceID. Although `Amplitude-Swift` doesn't support these configurations, you can add plugins to the new iOS SDK to enrich event payloads.

=== "Amplitude-Swift"

    ```swift

>     import AdSupportf

Reviewer 2: Typo `AdSupport`?

Author: Ha, a typo. Fixed.

---

The new version of Amplitude's iOS SDK (`Amplitude-Swift`) features a plugin architecture, built-in type definition and broader support for front-end frameworks. The new version isn't backwards compatible with `Amplitude-iOS`. 

To migrate to `Amplitude-Swift`, update your dependencies and instrumentation.

### Terminology

* `Amplitude-iOS`: Legacy iOS SDK
* `Amplitude-Swift`: New iOS SDK

## Dependency

For CocoaPods installation:
=== "Amplitude-iOS"
    Add dependency to Podfile.
    ```
    pod 'Amplitude', '~> 8.8.0'
    ```

=== "Amplitude-Swift"
    Add dependency to Podfile.
    ```
    pod 'AmplitudeSwift', '~> 0.3.0'
    ```

For Swift Package Manager installation:

=== "Amplitude-iOS"
    Enter `https://github.com/amplitude/Amplitude-iOS` into the search bar.

=== "Amplitude-Swift"
    Enter `https://github.com/amplitude/Amplitude-Swift` into the search bar.

For Carthage installation:

=== "Amplitude-iOS"
    Add the following line to your Cartfile.

    ```
    github "amplitude/Amplitude-iOS" ~> 8.8.0
    ```

=== "Amplitude-Swift"
    Add the following line to your Cartfile.

    ```

>     github "amplitude/Amplitude-Swift" ~> 8.8.0

Reviewer 3: version

    ```

## Instrumentation

The new iOS SDK offers an API to instrument events. To migrate to the new SDK, you need to update a few calls. The following sections detail which calls have changed.

### Initialization

> Like all other calls, `instance()` has been removed. legacy iOS and new iOS SDKs also has different way to set the configs. The new iOS SDKs intruduce Configuration object to set the configuration. See [Configuration](#configuration).

Reviewer 4: ```suggestion
Like all other calls, `instance()` has been removed. Configuration is handled differently between the legacy iOS and new iOS SDK. The new iOS SDKs use the Configuration object to set the configuration. See [Configuration](#configuration).
```

> Like all other calls, `instance()` has been removed. legacy iOS and new iOS SDKs also has different way to set the configs. The new iOS SDKs intruduce Configuration object to set the configuration. See [Configuration](#configuration).

Author: Previously the legacy SDK is used like `Amplitude.instace().xxx`. But the new SDK is used like `Amplitude(configuration: configuration).xxx`, instance() is not needed anymore.

> Like all other calls, `instance()` has been removed. legacy iOS and new iOS SDKs also has different way to set the configs. The new iOS SDKs intruduce Configuration object to set the configuration. See [Configuration](#configuration).

Author: All other calls mean `tack()`, `setUserId()` , `setDeviceId`, `setSessionId` etc. You can see the difference in the snippet. 
Let me know how to describe this in a better way. Thanks.

=== "Amplitude-iOS"

    ```swift
    import Amplitude

    Amplitude.instance().trackingSessionEvents = true
    Amplitude.instance().initializeApiKey("YOUR-API-KEY")
    ```

=== "Amplitude-Swift"

    ```swift
    import Amplitude_Swift

    Amplitude(
      configuration: Configuration(
        apiKey: "YOUR-API-KEY",
        trackingSessionEvents: true,
      )
    )
    ```

### Configuration

> The configuration new iOS SDK  comes in a different shape. The configurations are simpler and more consistent across runtimes.

Reviewer 4: ```suggestion
The configurations for the new SDK are simpler and more consistent across runtimes.
```

|`amplitude.eventUploadMaxBatchSize`|`config.flushQueueSize`|
|`amplitude.eventUploadThreshold`|`config.flushQueueSize`|
|`amplitude.eventUploadPeriodSeconds`|`config.flushIntervalMillis`|
|Set max retries count. NOT SUPPORTED.|`config.flushMaxRetries`|
|`amplitude.eventMaxCount`|NOT SUPPORTED.|
|`amplitude.optOut`|`config.optOut`|
|`amplitude.enableCoppaControl() or amplitude.disableCoppaControl()`|`config.enableCoppaControl`|
|Customize storage provider. NOT SUPPORTED.|`config.storageProvider`|
|Set up log level. NOT SUPPORTED.|`config.logLevel`|
|Customize logger provider. NOT SUPPORTED.|`config.loggerProvider`|
|Overwrite the minimum length deviceId and UserId. NOT SUPPORTED.|`config.minIdLength`|
|Partner Id for partner integrations. NOT SUPPORTED.|`config.partnerId`|
|The event callback. NOT SUPPORTED. See middleware. |`config.callback`|
|`amplitude.libraryName`|NOT SUPPORTED.|
|`amplitude.libraryVersion`|NOT SUPPORTED.|
|`amplitude.adSupportBlock`|NOT SUPPORTED. See Plugins.|
|`amplitude.useAdvertisingIdForDeviceId`|NOT SUPPORTED. See Plugins.|
|`amplitude.locationInfoBlock`|`amplitude.locationInfoBlock`|
|`amplitude.deferCheckInForeground`|NOT SUPPORTED.|

> |`amplitude.setOffline(Yes)`|NOT SUPPORT.|

Reviewer 4: ```suggestion
|`amplitude.setOffline(Yes)`|NOT SUPPORTED.|
```

> |`amplitude.setContentTypeHeader("YOUR-CONTENT-TYPE-HEADER")`|NOT SUPPORT.| 

Reviewer 4: ```suggestion
|`amplitude.setContentTypeHeader("YOUR-CONTENT-TYPE-HEADER")`|NOT SUPPORTED.| 
```

The `logEvent()` API maps to `track()`.

=== "Amplitude-iOS"

    ```swift
    let eventType = "Button Clicked"
    let eventProperties: [String: Any] = ["key": "value"]
    let groups: [String: Any] = ["orgId": 10]

    Amplitude.instance().logEvent(
        eventType,
        withEventProperties: eventProperties,
        withGroups: groups)
    ```

=== "Amplitude-Swift"

    ```swift
    let eventType = "Button Clicked"
    let event = BaseEvent(
      eventType: eventType, 
      eventProperties:[
        "integer": 1,
        "string": "stringValue",
        "array": [1, 2, 3],
      ], 
      groups: groups)
      
    amplitude.track(event)
    ```

#### `uploadEvents()`

The `uploadEvents()` API maps to `flush()`.

=== "Amplitude-iOS"

    ```swift
    Amplitude.instance().uploadEvents()
    ```

=== "Amplitude-Swift"

    ```swift
    amplitude.flush()
    ```

### Set user properties

The APIs for setting user properties are the same, except for the removal of `instance()`. Here are code snippets to migrate APIs for user properties.

#### `setUserId()`

Setting a user ID can be invoked on `amplitude` without calling `getInstance()`.

=== "Amplitude-iOS"

    ```swift

>     let userId = "TEST-USER-NAME"

Reviewer 4: Should "TEST-USER-NAME" be "TEST-USER-ID" or something like that?

    Amplitude.instance().setUserId(userId)
    ```

=== "Amplitude-Swift"

    ```swift
    let userId = "TEST-USER-NAME"
    amplitude.setUserId(userId: userId)
    ```

#### `setDeviceId()`

> Setting a device ID can be invoked on `amplitude` without calling `instance()`.

Reviewer 4: ```suggestion
Set a device ID on `amplitude` without calling `instance()`.
```

=== "Amplitude-iOS"

    ```swift
    let identify = AMPIdentify()
    identify.set("membership", value: "paid")
    Amplitude.instance().groupIdentify(
      withGroupType: "TEST-GROUP-TYPE", 
      groupName: "TEST-GROUP-NAME", 
      groupIdentify: identify
    )
    ```

=== "Amplitude-Swift"

    ```swift
    let identify = Identify()
    identify.set(property: "membership", value: "paid")
    amplitude.groupIdentify(
      groupType: "TEST-GROUP-TYPE", 
      groupName: "TEST-GROUP-NAME", 
      identify: identify
    )
    ```

### Tracking revenue

#### `logRevenueV2()`

Track revenue using `revenue()` API on `amplitude` without calling `instance()`.

=== "Amplitude-iOS"

    ```swift
      let revenue = AMPRevenue()
      revenue.setProductIdentifier("productIdentifier")
      revenue.setQuantity(3)
      revenue.setPrice(NSNumber(value: 3.99))

      Amplitude.instance().logRevenueV2(revenue)
    ```

=== "Amplitude-Swift"

    ```swift
    let revenue = Revenue()
    revenue.productId = "123"
    revenue1.price = 12
            
    amplitude.revenue(revenue: revenue)
    ```

### Patterns

#### Plugins

> The configs `amplitude.adSupportBlock` or `amplitude.useAdvertisingIdForDeviceId` were available in `Amplitude-iOS` to allow use idfv or idfa as the deviceID. Although `Amplitude-Swift` doesn't support these configurations, you can add plugins to the new iOS SDK to enrich event payloads.

Reviewer 4: ```suggestion
The configs `amplitude.adSupportBlock` or `amplitude.useAdvertisingIdForDeviceId` were available in `Amplitude-iOS` to allow you to use IDFV or IDFA as the deviceID. Although `Amplitude-Swift` doesn't support these configurations, you can add plugins to the new iOS SDK to enrich event payloads.
```

        weak var amplitude: Amplitude? = nil

        func execute(event: BaseEvent?) -> BaseEvent? {
            let status = ATTrackingManager.trackingAuthorizationStatus
            var idfa = fallbackValue
            if status == .authorized {
                idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
            }

            let workingEvent = event
            // The idfa on simulator is always 00000000-0000-0000-0000-000000000000
            event?.idfa = idfa
            // If you want to use idfa for the device_id
            event?.deviceId = idfa
            return workingEvent
        }
    }

    extension IDFACollectionPlugin {
        var fallbackValue: String? {
            // fallback to the IDFV value.
            // this is also sent in event.context.device.id,
            // feel free to use a value that is more useful to you.
            return UIDevice.current.identifierForVendor?.uuidString
        }
    }
    ```

To install your custom plugin, use `add()` with your custom plugin as parameter.

=== "Amplitude-Swift"

    ```swift
    amplitude.add(plugin: IDFACollectionPlugin())
    ```

#### Callback

> `Amplitude-Swft` supports configuration level and event level callback functions which will be called for success and error upload. Configuration level callback applies for every success and error event upload. Event level callback is specific for one Event. Notice that the event level callbacks are stored in cache. It will be loss when the app crash.

Reviewer 4: ```suggestion
`Amplitude-Swft` supports configuration-level and event-level callback functions which are called for success and error upload. Configuration-level callback applies for every success and error event upload. Event-level callback is specific for one Event. Notice that the event level callbacks are stored in cache. Event-level callbacks are lost if the app crashes.
```

=== "Amplitude-Swift"

    Configuration level callbacks:

    ```swift
    let amplitude = Amplitude(
        configuration: Configuration(
            apiKey: "TEST-API-KEY",
            callback: { (event: BaseEvent, code: Int, message: String) -> Void in
                print("eventcallback: \(event), code: \(code), message: \(message)")
            },
        )
    )
    ```

    Event level callbacks:

    ```swift
    let event = BaseEvent(
      callback: { (event: BaseEvent, code: Int, message: String) -> Void in
          print("eventcallback: \(event), code: \(code), message: \(message)")
      }, 
      eventType: "TEST-EVENT-TYPE")
      

>       amplitude.track(event: event)

Reviewer 2: Minor, indentation
